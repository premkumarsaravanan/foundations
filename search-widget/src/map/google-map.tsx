import React from 'react'
import styled from 'styled-components'
import GoogleMapLoader, { RenderProps } from './google-map-loader'
import { PropertyModel } from '../types/schema'
import { Params } from '../utils/query-params'

const MapDiv = styled.div`
  height: 500px;
  width: 500px;
`

const DEFAULT_CENTER = {lat: 51.507325, lng: -0.127716}
const DEFAULT_ZOOM = 8

export type GetContentParams = {
  latitude?: number
  longitude?: number
  address?: string
  qualifier?: string
  sellingPrice?: number
  bedrooms?: number
  bathrooms?: number
  marketingMode?: string
  lettingPrice?: number
  rentFrequency?: string
}

export const getContent = ({ latitude, longitude, address, qualifier, sellingPrice, bedrooms, bathrooms, marketingMode , lettingPrice, rentFrequency }: GetContentParams) => `
  <div style="display:flex;" id="coordinate-${latitude}-${longitude}">
    <div><img src="https://pix5.agoda.net/hotelImages/773/77384/77384_14091113470022048575.jpg?s=100x100"></div>
    <div style="padding: 0rem 1rem;">
      <div style="font-weight:bold;font-size:1rem;">${address}</div>
      ${marketingMode === 'selling' ? (
        `<!-- <div style="color: #00a569;font-weight:bold;font-size:2rem">${qualifier}</div> -->
        <div style="color: #00a569;font-weight:bold;font-size:2rem">£${sellingPrice}</div>`
      ) : (
        `<!-- <div style="color: #00a569;font-weight:bold;font-size:2rem">${rentFrequency}</div> -->
        <div style="color: #00a569;font-weight:bold;font-size:2rem">£${lettingPrice}</div>`
      )}
      <div style="display:flex;justify-content: space-around;">
        <div>
          <span>
            <svg aria-hidden="true" focusable="false" data-prefix="fas" data-icon="bed" class="svg-inline--fa fa-bed fa-w-20" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 640 512">
              <path fill="currentColor" d="M176 256c44.11 0 80-35.89 80-80s-35.89-80-80-80-80 35.89-80 80 35.89 80 80 80zm352-128H304c-8.84 0-16 7.16-16 16v144H64V80c0-8.84-7.16-16-16-16H16C7.16 64 0 71.16 0 80v352c0 8.84 7.16 16 16 16h32c8.84 0 16-7.16 16-16v-48h512v48c0 8.84 7.16 16 16 16h32c8.84 0 16-7.16 16-16V240c0-61.86-50.14-112-112-112z">
              </path>
            </svg>
          </span>
          <span>${bedrooms}</span>
        </div>
        <div>
          <span>
            <svg aria-hidden="true" focusable="false" data-prefix="fas" data-icon="bath" class="svg-inline--fa fa-bath fa-w-16" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512">
              <path fill="currentColor" d="M488 256H80V112c0-17.645 14.355-32 32-32 11.351 0 21.332 5.945 27.015 14.88-16.492 25.207-14.687 59.576 6.838 83.035-4.176 4.713-4.021 11.916.491 16.428l11.314 11.314c4.686 4.686 12.284 4.686 16.971 0l95.03-95.029c4.686-4.686 4.686-12.284 0-16.971l-11.314-11.314c-4.512-4.512-11.715-4.666-16.428-.491-17.949-16.469-42.294-21.429-64.178-15.365C163.281 45.667 139.212 32 112 32c-44.112 0-80 35.888-80 80v144h-8c-13.255 0-24 10.745-24 24v16c0 13.255 10.745 24 24 24h8v32c0 28.43 12.362 53.969 32 71.547V456c0 13.255 10.745 24 24 24h16c13.255 0 24-10.745 24-24v-8h256v8c0 13.255 10.745 24 24 24h16c13.255 0 24-10.745 24-24v-32.453c19.638-17.578 32-43.117 32-71.547v-32h8c13.255 0 24-10.745 24-24v-16c0-13.255-10.745-24-24-24z">
              </path>
            </svg>
          </span>
          <span>${bathrooms}</span>
        </div>
      </div>
    </div>
  </div>
`

export type HandleUseEffectParams = {
  googleMap: any
  center?: GeoLocation
  zoom?: number
  property?: PropertyModel
  mapRef: any
  restProps?: any
}

export const handleUseEffect = ({ googleMap, center, zoom, restProps, property, mapRef }: HandleUseEffectParams) => () => {
  if(googleMap) {
    const map = new googleMap.Map(document.getElementById('map-container'), {
      center: center || DEFAULT_CENTER,
      zoom: zoom || DEFAULT_ZOOM,
      styles: [
        {
            "featureType": "all",
            "elementType": "geometry.fill",
            "stylers": [
                {
                    "weight": "2.00"
                }
            ]
        },
        {
            "featureType": "all",
            "elementType": "geometry.stroke",
            "stylers": [
                {
                    "color": "#9c9c9c"
                }
            ]
        },
        {
            "featureType": "all",
            "elementType": "labels.text",
            "stylers": [
                {
                    "visibility": "on"
                }
            ]
        },
        {
            "featureType": "landscape",
            "elementType": "all",
            "stylers": [
                {
                    "color": "#f2f2f2"
                }
            ]
        },
        {
            "featureType": "landscape",
            "elementType": "geometry.fill",
            "stylers": [
                {
                    "color": "#ffffff"
                }
            ]
        },
        {
            "featureType": "landscape.man_made",
            "elementType": "geometry.fill",
            "stylers": [
                {
                    "color": "#ffffff"
                }
            ]
        },
        {
            "featureType": "poi",
            "elementType": "all",
            "stylers": [
                {
                    "visibility": "off"
                }
            ]
        },
        {
            "featureType": "road",
            "elementType": "all",
            "stylers": [
                {
                    "saturation": -100
                },
                {
                    "lightness": 45
                }
            ]
        },
        {
            "featureType": "road",
            "elementType": "geometry.fill",
            "stylers": [
                {
                    "color": "#eeeeee"
                }
            ]
        },
        {
            "featureType": "road",
            "elementType": "labels.text.fill",
            "stylers": [
                {
                    "color": "#7b7b7b"
                }
            ]
        },
        {
            "featureType": "road",
            "elementType": "labels.text.stroke",
            "stylers": [
                {
                    "color": "#ffffff"
                }
            ]
        },
        {
            "featureType": "road.highway",
            "elementType": "all",
            "stylers": [
                {
                    "visibility": "simplified"
                }
            ]
        },
        {
            "featureType": "road.arterial",
            "elementType": "labels.icon",
            "stylers": [
                {
                    "visibility": "off"
                }
            ]
        },
        {
            "featureType": "transit",
            "elementType": "all",
            "stylers": [
                {
                    "visibility": "off"
                }
            ]
        },
        {
            "featureType": "water",
            "elementType": "all",
            "stylers": [
                {
                    "color": "#46bcec"
                },
                {
                    "visibility": "on"
                }
            ]
        },
        {
            "featureType": "water",
            "elementType": "geometry.fill",
            "stylers": [
                {
                    "color": "#c8d7d4"
                }
            ]
        },
        {
            "featureType": "water",
            "elementType": "labels.text.fill",
            "stylers": [
                {
                    "color": "#070707"
                }
            ]
        },
        {
            "featureType": "water",
            "elementType": "labels.text.stroke",
            "stylers": [
                {
                    "color": "#ffffff"
                }
            ]
          }
        ],
      ...restProps
    })
    const latitude = property && property.address && property.address.geolocation && property.address.geolocation.latitude
    const longitude = property && property.address && property.address.geolocation && property.address.geolocation.longitude
    const marketingMode = property && property.marketingMode

    mapRef.current = map
    const marker = new googleMap.Marker({
      position: {lat: latitude || DEFAULT_CENTER.lat, lng: longitude|| DEFAULT_CENTER.lng},
      icon: null,
      map
    })
    const address = property && property.address && property.address.line2 ?  property.address.line2 : ''
    const qualifier = property && property.selling && property.selling.qualifier
    const sellingPrice = property && property.selling && property.selling.price

    const lettingPrice = property && property.letting && property.letting.rent
    const rentFrequency = property && property.letting && property.letting.rentFrequency

    const bedrooms = property && property.bedrooms
    const bathrooms = property && property.bathrooms
    const infoWindow = new googleMap.InfoWindow({
      content: getContent({ latitude, longitude, qualifier, sellingPrice, bedrooms, bathrooms, address, marketingMode, lettingPrice, rentFrequency })
    })
    infoWindow.open(map, marker)
  }
}

export type MapContainerProps = {
  googleMap: any
  center?: GeoLocation
  zoom?: number
  property?: PropertyModel
}

export const MapContainer:React.FC<MapContainerProps> = ({ googleMap, center, zoom, property, ...restProps }) => {
  const mapRef = React.useRef(null)
  React.useEffect(handleUseEffect({googleMap, zoom, center, mapRef, restProps, property}), [googleMap, restProps, center, zoom, property])
  return <MapDiv id="map-container"/>
}

export type RenderMapParams = {
  property?: PropertyModel
  zoom?: number
  center?: GeoLocation
}

export const renderMap = ({ property, zoom, center }: RenderMapParams) => ({ googleMap, error }: RenderProps) => {
  if(error) {
    return <div>{error}</div>
  }
return <MapContainer googleMap={googleMap} property={property} zoom={zoom} center={center} />
}

export type GeoLocation = {
  lat: number
  lng: number
}

export type GoogleMapProps = {
  params: Params
  property?: PropertyModel
  zoom?: number
  center?: GeoLocation
}

export const GoogleMap: React.FC<GoogleMapProps> = ({ params, property, zoom, center }) => {
  return(
    <GoogleMapLoader
      params={params}
      render={renderMap({ property, zoom, center})}
    />
  )
}

export default GoogleMap