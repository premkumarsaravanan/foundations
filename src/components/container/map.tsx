import * as React from 'react'
import { CoordinateProps, Coords, Map, combineAddress } from '@reapit/elements'
import { oc } from 'ts-optchain'
import { connect } from 'react-redux'
import { ReduxState } from '@/types/core'
import { AppointmentModel, AddressModel } from '@/types/appointments'
import invalidValues from '@/constants/invalid-values'
import styles from '@/styles/pages/map.scss?mod'
import MapPanel from '../ui/map-panel'
import { appointmentDetailRequestData } from '@/actions/appointment-detail'
import { Dispatch } from 'redux'
import { mapStyles } from './map-styles'

const { UNDEFINED_LATLNG_NUMBER, UNDEFINED_NULL_STRING } = invalidValues

export interface MapContainerMappedState {
  appointments: AppointmentModel[] | undefined
  destinationLatLng: {
    lat: number | undefined
    lng: number | undefined
  }
  destinationAddress?: string
  desktopMode: boolean
}

export interface Coordinate {
  position: {
    lat: number
    lng: number
  }
  address: AddressModel
  id: string
}

export type MapContainerMappedActions = {
  handleOnClick: (id: string) => () => void
}

export type MapContainerProps = MapContainerMappedState &
  MapContainerMappedActions & {
    travelMode: 'DRIVING' | 'WALKING'
  }

export const filterInvalidMarker = (markers: Coordinate[]) => {
  return markers.filter(appointments => {
    return (
      appointments.position.lat !== UNDEFINED_LATLNG_NUMBER &&
      appointments.position.lng !== UNDEFINED_LATLNG_NUMBER &&
      appointments.id !== UNDEFINED_NULL_STRING
    )
  })
}

export const MapContainer = ({
  appointments = [],
  destinationLatLng,
  destinationAddress,
  travelMode,
  handleOnClick,
  desktopMode
}: MapContainerProps) => {
  const [distance, setDistance] = React.useState('')
  const [duration, setDuration] = React.useState('')
  const [currentLocation, setCurrentLocation] = React.useState<Coords>({
    lat: UNDEFINED_LATLNG_NUMBER,
    lng: UNDEFINED_LATLNG_NUMBER
  })

  const coordinates: CoordinateProps<any> = filterInvalidMarker(
    appointments.map(
      (appointment: AppointmentModel): Coordinate => {
        const lat = oc(appointment).property.address.geolocation.latitude(UNDEFINED_LATLNG_NUMBER)
        const lng = oc(appointment).property.address.geolocation.longitude(UNDEFINED_LATLNG_NUMBER)
        const id = oc(appointment).id(UNDEFINED_NULL_STRING)
        const address = oc(appointment).property.address({})
        return {
          id,
          address,
          position: {
            lat,
            lng
          }
        }
      }
    )
  )

  React.useEffect(() => {
    if (navigator.geolocation) {
      navigator.geolocation.getCurrentPosition(position => {
        setCurrentLocation({ ...currentLocation, lat: position.coords.latitude, lng: position.coords.longitude })
      })
    }
  }, [])

  const onLoadedDirection = React.useCallback(
    res => {
      const { duration, distance } = res.routes[0].legs[0]
      setDistance(distance.text)
      setDuration(duration.text)
    },
    [destinationLatLng]
  )

  const hasMapPanel = destinationLatLng.lng && destinationLatLng.lat

  return (
    <>
      <div
        className={`${styles.mapContainer} ${desktopMode ? styles.isDesktop : ''} ${
          hasMapPanel ? styles.mapHasPanel : ''
        }`}
      >
        <Map
          autoFitBounds={true}
          apiKey={process.env.MAP_API_KEY as string}
          coordinates={coordinates}
          markerCallBack={handleOnClick}
          destinationPoint={destinationLatLng}
          destinationAddress={destinationAddress}
          travelMode={travelMode}
          onLoadedDirection={onLoadedDirection}
          mapContainerStyles={{ height: '100%' }}
          styles={mapStyles}
        />
      </div>
      <MapPanel
        duration={duration}
        distance={distance}
        currentLocation={currentLocation}
        destination={destinationLatLng}
      />
    </>
  )
}

export const mapStateToProps = (state: ReduxState): MapContainerMappedState => {
  const appointments = oc(state).appointments.appointments.data([])
  const destinationLat = oc(state).direction.destination.property.address.geolocation.latitude(undefined)
  const destinationLng = oc(state).direction.destination.property.address.geolocation.longitude(undefined)
  const destinationAddress = combineAddress(oc(state).direction.destination.property.address({}))
  return {
    appointments,
    destinationLatLng: {
      lat: destinationLat,
      lng: destinationLng
    },
    destinationAddress,
    desktopMode: oc(state).auth.refreshSession.mode() === 'DESKTOP'
  }
}

export const mapDispatchToProps = (dispatch: Dispatch): MapContainerMappedActions => ({
  handleOnClick: (id: string) => () => {
    dispatch(appointmentDetailRequestData({ id }))
  }
})

const MapContainerWithConnect = connect(
  mapStateToProps,
  mapDispatchToProps
)(MapContainer)

MapContainerWithConnect.displayName = 'CurrentLocButtonWithConnect'

export default MapContainerWithConnect
