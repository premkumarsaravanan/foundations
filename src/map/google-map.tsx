import React from 'react'
import { INVALID_BACKGROUND_AS_BASE64 } from '../constants'
import { context } from '../context'
import styled from 'styled-components'
import GoogleMapLoader, { RenderProps } from './google-map-loader'
import { PropertyModel } from '../types/property'
import { Params } from '../utils/query-params'
import { Theme } from '../theme'
import { getPrice } from '../search-result'

const { useContext } = React

const MapDiv = styled.div`
  min-height: 500px;
  @media screen and (max-width: 1600px) {
    & {
      height: 500px;
    }
  }

  width: 100%;
  height: 100%;
`

const DEFAULT_CENTER = { lat: 51.507325, lng: -0.127716 }
const DEFAULT_ZOOM = 15

export type GetContentParams = {
  latitude?: number
  longitude?: number
  address?: string
  bedrooms?: number
  bathrooms?: number
  marketingMode?: string
  lettingPrice?: number
  rentFrequency?: string
  imageUrl?: string
  theme: Theme
  price: string
}

export const getContent = ({
  price,
  theme,
  latitude,
  longitude,
  address,
  bedrooms,
  bathrooms,
  marketingMode,
  imageUrl = INVALID_BACKGROUND_AS_BASE64
}: GetContentParams) => `
  <div style="display:flex; font-family: ${
    theme.base.font.family
  }" id="coordinate-${latitude}-${longitude}">
    <div><img style="width: 110px; height: 110px; object-fit: cover" src="${imageUrl}"></div>
    <div style="padding: 0rem 1rem;">
      <div style="margin-bottom: 2px; font-weight:bold;font-size:1rem; color: ${
        theme.colors.base
      }">${address}</div>
      ${
        marketingMode === 'selling'
          ? `<div style="color: ${theme.colors.primary};font-weight:bold;font-size:2rem">£${price}</div>`
          : `<div style="color: ${theme.colors.primary};font-weight:bold;font-size:2rem">£${price}</div>`
      }
      <div style="display:flex; margin-top: 2px">
        <div style="margin-right: 1.2rem">
          <span style="color: ${theme.colors.icon}">
            <svg style="width:20px; height:20px" aria-hidden="true" focusable="false" data-prefix="fas" data-icon="bed" class="svg-inline--fa fa-bed fa-w-20" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 640 512">
              <path fill="currentColor" d="M176 256c44.11 0 80-35.89 80-80s-35.89-80-80-80-80 35.89-80 80 35.89 80 80 80zm352-128H304c-8.84 0-16 7.16-16 16v144H64V80c0-8.84-7.16-16-16-16H16C7.16 64 0 71.16 0 80v352c0 8.84 7.16 16 16 16h32c8.84 0 16-7.16 16-16v-48h512v48c0 8.84 7.16 16 16 16h32c8.84 0 16-7.16 16-16V240c0-61.86-50.14-112-112-112z">
              </path>
            </svg>
          </span>
          <div>${bedrooms}</div>
        </div>
        <div>
          <span style="color: ${theme.colors.icon}">
            <svg style="width:20px; height:20px" aria-hidden="true" focusable="false" data-prefix="fas" data-icon="bath" class="svg-inline--fa fa-bath fa-w-16" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512">
              <path fill="currentColor" d="M488 256H80V112c0-17.645 14.355-32 32-32 11.351 0 21.332 5.945 27.015 14.88-16.492 25.207-14.687 59.576 6.838 83.035-4.176 4.713-4.021 11.916.491 16.428l11.314 11.314c4.686 4.686 12.284 4.686 16.971 0l95.03-95.029c4.686-4.686 4.686-12.284 0-16.971l-11.314-11.314c-4.512-4.512-11.715-4.666-16.428-.491-17.949-16.469-42.294-21.429-64.178-15.365C163.281 45.667 139.212 32 112 32c-44.112 0-80 35.888-80 80v144h-8c-13.255 0-24 10.745-24 24v16c0 13.255 10.745 24 24 24h8v32c0 28.43 12.362 53.969 32 71.547V456c0 13.255 10.745 24 24 24h16c13.255 0 24-10.745 24-24v-8h256v8c0 13.255 10.745 24 24 24h16c13.255 0 24-10.745 24-24v-32.453c19.638-17.578 32-43.117 32-71.547v-32h8c13.255 0 24-10.745 24-24v-16c0-13.255-10.745-24-24-24z">
              </path>
            </svg>
          </span>
          <div>${bathrooms}</div>
        </div>
      </div>
    </div>
  </div>
`

export type HandleUseEffectParams = {
  googleMap: any
  center?: GeoLocation
  zoom?: number
  property?: PropertyModel
  mapRef: any
  restProps?: any
  imageUrl: string | undefined
  theme: Theme | null
  price: string
}

export const handleUseEffect = ({
  price,
  googleMap,
  center,
  zoom,
  restProps,
  property,
  mapRef,
  imageUrl,
  theme
}: HandleUseEffectParams) => () => {
  if (googleMap && theme) {
    const map = new googleMap.Map(
      document.getElementById('reapit-map-container'),
      {
        center: center || DEFAULT_CENTER,
        zoom: zoom || DEFAULT_ZOOM,
        styles: [
          {
            featureType: 'all',
            elementType: 'geometry.fill',
            stylers: [
              {
                weight: '2.00'
              }
            ]
          },
          {
            featureType: 'all',
            elementType: 'geometry.stroke',
            stylers: [
              {
                color: '#9c9c9c'
              }
            ]
          },
          {
            featureType: 'all',
            elementType: 'labels.text',
            stylers: [
              {
                visibility: 'on'
              }
            ]
          },
          {
            featureType: 'landscape',
            elementType: 'all',
            stylers: [
              {
                color: '#f2f2f2'
              }
            ]
          },
          {
            featureType: 'landscape',
            elementType: 'geometry.fill',
            stylers: [
              {
                color: '#ffffff'
              }
            ]
          },
          {
            featureType: 'landscape.man_made',
            elementType: 'geometry.fill',
            stylers: [
              {
                color: '#ffffff'
              }
            ]
          },
          {
            featureType: 'poi',
            elementType: 'all',
            stylers: [
              {
                visibility: 'off'
              }
            ]
          },
          {
            featureType: 'road',
            elementType: 'all',
            stylers: [
              {
                saturation: -100
              },
              {
                lightness: 45
              }
            ]
          },
          {
            featureType: 'road',
            elementType: 'geometry.fill',
            stylers: [
              {
                color: '#eeeeee'
              }
            ]
          },
          {
            featureType: 'road',
            elementType: 'labels.text.fill',
            stylers: [
              {
                color: '#7b7b7b'
              }
            ]
          },
          {
            featureType: 'road',
            elementType: 'labels.text.stroke',
            stylers: [
              {
                color: '#ffffff'
              }
            ]
          },
          {
            featureType: 'road.highway',
            elementType: 'all',
            stylers: [
              {
                visibility: 'simplified'
              }
            ]
          },
          {
            featureType: 'road.arterial',
            elementType: 'labels.icon',
            stylers: [
              {
                visibility: 'off'
              }
            ]
          },
          {
            featureType: 'transit',
            elementType: 'all',
            stylers: [
              {
                visibility: 'off'
              }
            ]
          },
          {
            featureType: 'water',
            elementType: 'all',
            stylers: [
              {
                color: '#46bcec'
              },
              {
                visibility: 'on'
              }
            ]
          },
          {
            featureType: 'water',
            elementType: 'geometry.fill',
            stylers: [
              {
                color: '#c8d7d4'
              }
            ]
          },
          {
            featureType: 'water',
            elementType: 'labels.text.fill',
            stylers: [
              {
                color: '#070707'
              }
            ]
          },
          {
            featureType: 'water',
            elementType: 'labels.text.stroke',
            stylers: [
              {
                color: '#ffffff'
              }
            ]
          }
        ],
        ...restProps
      }
    )
    const latitude =
      property &&
      property.address &&
      property.address.geolocation &&
      property.address.geolocation.latitude
    const longitude =
      property &&
      property.address &&
      property.address.geolocation &&
      property.address.geolocation.longitude
    const marketingMode = property && property.marketingMode

    mapRef.current = map

    if (property) {
      const marker = new googleMap.Marker({
        position: {
          lat: latitude || DEFAULT_CENTER.lat,
          lng: longitude || DEFAULT_CENTER.lng
        },
        icon: null,
        map
      })
      const address =
        property && property.address && property.address.line2
          ? property.address.line2
          : ''

      const lettingPrice = property && property.letting && property.letting.rent
      const rentFrequency =
        property && property.letting && property.letting.rentFrequency

      const bedrooms = property && property.bedrooms
      const bathrooms = property && property.bathrooms
      const infoWindow = new googleMap.InfoWindow({
        content: getContent({
          price,
          theme,
          latitude,
          longitude,
          bedrooms,
          bathrooms,
          address,
          marketingMode,
          lettingPrice,
          rentFrequency,
          imageUrl
        })
      })
      infoWindow.open(map, marker)
    }
  }
}

export type MapContainerProps = {
  googleMap: any
  center?: GeoLocation
  zoom?: number
  property?: PropertyModel
}

export const MapContainer: React.FC<MapContainerProps> = ({
  googleMap,
  center,
  zoom,
  property,
  ...restProps
}) => {
  const mapRef = React.useRef(null)
  const contextValue = useContext(context)
  const searchStore = useContext(context)

  let price = ''
  if (searchStore && searchStore.selectedProperty) {
    price = getPrice(searchStore.selectedProperty, searchStore.searchType)
  }

  let imageUrl = INVALID_BACKGROUND_AS_BASE64
  if (searchStore && property) {
    const propertyId = property && property.id
    const propertyImage = propertyId
      ? searchStore.propertyImages[propertyId]
      : ''
    if (propertyImage && propertyImage.url) {
      imageUrl = propertyImage.url
    }
  }

  React.useEffect(
    handleUseEffect({
      price,
      googleMap,
      zoom,
      center,
      mapRef,
      restProps,
      property,
      imageUrl,
      theme: contextValue ? contextValue.theme : null
    }),
    [googleMap, center, zoom, property]
  )

  if (!contextValue) {
    return null
  }
  return <MapDiv id="reapit-map-container"></MapDiv>
}

export type RenderMapParams = {
  property?: PropertyModel
  zoom?: number
  center?: GeoLocation
}

export const renderMap = ({ property, zoom, center }: RenderMapParams) => ({
  googleMap,
  error
}: RenderProps) => {
  if (error) {
    return <div>{error}</div>
  }
  return (
    <MapContainer
      googleMap={googleMap}
      property={property}
      zoom={zoom}
      center={center}
    />
  )
}

export type GeoLocation = {
  lat: number
  lng: number
}

export type GoogleMapProps = {
  params: Params
  property?: PropertyModel
  zoom?: number
  center?: GeoLocation
}

export const GoogleMap: React.FC<GoogleMapProps> = ({
  params,
  property,
  zoom,
  center
}) => {
  return (
    <GoogleMapLoader
      params={params}
      render={renderMap({ property, zoom, center })}
    />
  )
}

export default GoogleMap
