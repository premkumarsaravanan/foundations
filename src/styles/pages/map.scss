@import '../base/colors.scss';

.mapContainer {
  overflow: hidden;
  position: absolute;
  left: 300px;
  width: calc(100vw - 300px);
  top: 73px;
  height: calc(100vh - 73px);

  @media screen and (max-width: 768px) {
    left: 0;
    top: 108px;
    height: calc(100vh - 108px);
    width: 100%;
  }
}

.mapContainer.mapHasPanel {
  height: calc(100vh - 165px);

  @media screen and (max-width: 768px) {
    height: calc(100vh - 200px);
  }
}

.mapPanel {
  position: absolute;
  bottom: 0;
  width: calc(100vw - 300px);
  left: 300px;
  display: flex;
  background-color: #fff;
  padding: 1.5rem;
  & > * {
    margin-right: 10px;
  }
  @media screen and (max-width: 768px) {
    justify-content: space-between;
    & > * {
      margin-right: 0;
    }
    left: 0;
    width: 100%;
  }
}
