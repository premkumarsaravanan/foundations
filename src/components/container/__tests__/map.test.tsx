import * as React from 'react'
import invalidValues from '@/constants/invalid-values'
import { MapContainer, filterInvalidMarker } from '../map'
import { shallow } from 'enzyme'

const { UNDEFINED_LATLNG_NUMBER } = invalidValues

describe('Map', () => {
  it('filter invalid marker correctly', () => {
    const input = [
      {
        lat: UNDEFINED_LATLNG_NUMBER,
        lng: UNDEFINED_LATLNG_NUMBER,
        title: 'not ok',
        content: 'not ok'
      },
      {
        lat: 0,
        lng: 0,
        title: 'ok',
        content: 'ok'
      }
    ]

    expect(filterInvalidMarker(input)).toEqual([
      {
        lat: 0,
        lng: 0,
        title: 'ok',
        content: 'ok'
      }
    ])
  })
  it('Should match snapshot', () => {
    expect(shallow(<MapContainer appointments={[]} />)).toMatchSnapshot()
  })
})
