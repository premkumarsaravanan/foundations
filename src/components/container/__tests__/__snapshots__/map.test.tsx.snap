// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map Should match snapshot 1`] = `
<div>
  <Map
    autoFitBounds={true}
    component={
      Object {
        "$$typeof": Symbol(react.memo),
        "WrappedComponent": [Function],
        "compare": null,
        "displayName": "Connect(Component)",
        "type": [Function],
      }
    }
    coordinates={
      Array [
        Object {
          "address1": "High Street",
          "address2": "The Stables",
          "id": "NEP1600290",
          "position": Object {
            "lat": 52.079532,
            "lng": -0.790871,
          },
        },
        Object {
          "address1": "The Embankment",
          "address2": "Bedford",
          "id": "BED1600537",
          "position": Object {
            "lat": 52.135071,
            "lng": -0.45936,
          },
        },
        Object {
          "address1": "Sandhill Road",
          "address2": "East Claydon",
          "id": "BUC1600211",
          "position": Object {
            "lat": 51.924759,
            "lng": -0.929362,
          },
        },
        Object {
          "address1": "Little Tingewick",
          "address2": "Buckingham",
          "id": "BUC1600210",
          "position": Object {
            "lat": 51.989592,
            "lng": -1.06774,
          },
        },
        Object {
          "address1": "",
          "address2": "",
          "id": "STS1600403",
          "position": Object {
            "lat": 9999,
            "lng": 9999,
          },
        },
        Object {
          "address1": "",
          "address2": "",
          "id": "STS1600401",
          "position": Object {
            "lat": 9999,
            "lng": 9999,
          },
        },
        Object {
          "address1": "High Street",
          "address2": "The Stables",
          "id": "NEP1600288",
          "position": Object {
            "lat": 52.079532,
            "lng": -0.790871,
          },
        },
        Object {
          "address1": "",
          "address2": "",
          "id": "STS1600396",
          "position": Object {
            "lat": 9999,
            "lng": 9999,
          },
        },
        Object {
          "address1": "",
          "address2": "",
          "id": "NEP1600287",
          "position": Object {
            "lat": 9999,
            "lng": 9999,
          },
        },
        Object {
          "address1": "Shelton Road",
          "address2": "Yielden",
          "id": "BED1600534",
          "position": Object {
            "lat": 52.29491,
            "lng": -0.512901,
          },
        },
        Object {
          "address1": "",
          "address2": "",
          "id": "STS1600394",
          "position": Object {
            "lat": 9999,
            "lng": 9999,
          },
        },
        Object {
          "address1": "The Causeway",
          "address2": "Carlton",
          "id": "BED1600532",
          "position": Object {
            "lat": 52.188155,
            "lng": -0.604618,
          },
        },
        Object {
          "address1": "Foxfield",
          "address2": "Broughton Gate",
          "id": "MKC1600312",
          "position": Object {
            "lat": 52.044516,
            "lng": -0.689953,
          },
        },
        Object {
          "address1": "Bedford Road",
          "address2": "Northill",
          "id": "BED1600530",
          "position": Object {
            "lat": 52.105969,
            "lng": -0.353784,
          },
        },
        Object {
          "address1": "Little Tingewick",
          "address2": "Buckingham",
          "id": "BUC1600209",
          "position": Object {
            "lat": 51.989592,
            "lng": -1.06774,
          },
        },
        Object {
          "address1": "The Embankment",
          "address2": "Bedford",
          "id": "BED1600525",
          "position": Object {
            "lat": 52.135071,
            "lng": -0.45936,
          },
        },
        Object {
          "address1": "Tyne Crescent",
          "address2": "Bedford",
          "id": "BED1600523",
          "position": Object {
            "lat": 52.161384,
            "lng": -0.461113,
          },
        },
        Object {
          "address1": "Deacon Place",
          "address2": "Middleton",
          "id": "MKC1600311",
          "position": Object {
            "lat": 52.04681,
            "lng": -0.710721,
          },
        },
        Object {
          "address1": "",
          "address2": "",
          "id": "STS1600389",
          "position": Object {
            "lat": 9999,
            "lng": 9999,
          },
        },
      ]
    }
    defaultZoom={16}
    destinationPoint={
      Object {
        "lat": 0,
        "lng": 0,
      }
    }
    travelMode="DRIVING"
  />
  <Memo()
    onChangeTravelMode={[Function]}
    travelMode="DRIVING"
  />
</div>
`;

exports[`Map Should match snapshot 2`] = `
<div>
  <Map
    autoFitBounds={true}
    component={
      Object {
        "$$typeof": Symbol(react.memo),
        "WrappedComponent": [Function],
        "compare": null,
        "displayName": "Connect(Component)",
        "type": [Function],
      }
    }
    coordinates={Array []}
    defaultZoom={16}
    destinationPoint={
      Object {
        "lat": 0,
        "lng": 0,
      }
    }
    travelMode="DRIVING"
  />
  <Memo()
    onChangeTravelMode={[Function]}
    travelMode="DRIVING"
  />
</div>
`;
